src/tracer.c

```diff
@@ -186,42 +215,22 @@ handle_syscall(pid_t pid, const struct ptrace_syscall_info *entry,
 	return;			       // return on syscall failure
     }
 
-    const int syscall = entry->entry.nr;
-
-    if (syscall != SYS_open && syscall != SYS_creat && syscall != SYS_openat
-	&& syscall != SYS_close) {
-	return;			       // return if we don't care about
-	// tracking said syscall
-    }
-
-    if (syscall == SYS_close) {
-	handle_close(pinfo->finfo + entry->entry.args[0]);
-	return;
-    }
-
-    const int fd = exit->exit.rval;
-
-    if (fd >= 1024)		       // more than 1024 files open
-	// concurrently
-    {
-	error(EXIT_FAILURE, errno,
-	      "limit of 1024 open files exceeded for process %d", pid);
+    switch (entry->entry.nr) {
+	case SYS_open:
+	    handle_open(pid, entry->entry.args, exit->exit.rval);
+	    break;
+	case SYS_creat:
+	    handle_creat(pid, (void *) entry->entry.args[0], exit->exit.rval);
+	    break;
+	case SYS_openat:
+	    handle_openat(pid, entry->entry.args, exit->exit.rval);
+	    break;
+	case SYS_close:
+	    handle_close(pid, (int) entry->entry.args[0]);
+	    break;
+	default:
+	    return;
```

Comment:
Do not hide information via parameter passing (e.g. pass exit.rval, expect fd) and do not use so different function signatures (e.g. args or args[0]).

Use local/temp variables to make it clear, e.g.
`int fd = exit->exit.rval;`

Since all `handle` functions are a couple of lines, let's start by having all the code here in the switch -- we can then see whether it makes sense to split it.